### PR TITLE
Websocket client

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: fluence-js/src/WebsocketSession.ts [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+        const tx = prepareRequest(payload, this.sessionId, counter, this.privateKey);
+
+        const request = {
+            tx: tx.payload,
+            request_id: requestId,
+            type: "tx_request"
+        };
+
+        return this.sendAndWaitResponse(requestId, JSON.stringify(request)).then((r) => {});
```
Comment (folex) [inline fluence-js/src/WebsocketSession.ts]: `.then((r) => {})`?

File: fluence-js/src/WebsocketSession.ts [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+        } catch (e) {
+            console.error("Cannot parse websocket event: " + e)
+        }
+    }
+
+    /**
+     * Trying to subscribe to all existed subscriptions again.
+     */
+    private resubscribe() {
```
[the remark this answers is not in the document]
Comment (DieMyst) [reply fluence-js/src/WebsocketSession.ts]: fixes

File: fluence-js/src/executor.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Executor based on promise. Terminates once.
+ */
+export class PromiseExecutor<T> extends Executor<T> {
+    private resultResolver: (result: T) => void;
+    private errorResolver: (error: any) => void;
+    readonly promise: Promise<T>;
+    readonly creationTime: number;
+    private timeout: ReturnType<typeof setTimeout> | undefined;
```
Comment (folex) [inline fluence-js/src/executor.ts]: Why do you need `ReturnType` here? Is it a timeout, like in seconds or something?

Comment (DieMyst) [reply fluence-js/src/executor.ts]: there is no `Timeout` type in typescript. It is a hack to use some type, that returns from `setTimeout`

File: fluence-js/src/WebsocketSession.ts [hunk 4]
```diff
@@ -282,31 +177,218 @@ export class WebsocketSession {
...
+        debug(`Sending message ${message}`);
         this.socket.send(message);
 
-        const timeout = setTimeout(() => {
+        // if timeout occurred, delete executor from state and do reconnect
+        const onTimeout = () => {
             if (this.executors.has(requestId)) {
-                executor.handleError(`Timeout after ${this.timeout} milliseconds.`);
+                executor.fail(`Timeout after ${this.timeout} milliseconds.`);
                 this.executors.delete(requestId);
```
Comment (folex) [inline fluence-js/src/WebsocketSession.ts]: Will this also delete subscriptions?

**Pull request merged**